dnf/transaction_sr.py

```diff
@@ -0,0 +1,104 @@
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Library General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+
+from __future__ import absolute_import
+from __future__ import print_function
+from __future__ import unicode_literals
+
+import libdnf
+
+from dnf.i18n import _
+import dnf.exceptions
+
+import json
+
+
+VERSION_MAJOR = "1"
+VERSION_MINOR = "0"
+VERSION = "%s.%s" % (VERSION_MAJOR, VERSION_MINOR)
+
+
+class StoredTransactionVersionError(dnf.exceptions.Error):
+    def __init__(self, value=None):
+        super(StoredTransactionVersionError, self).__init__(value)
+
+
+def check_version(version, filename):
+    major, minor = version.split('.')
+    if major != VERSION_MAJOR:
+        raise StoredTransactionVersionError(
+            _('Incompatible major version {major} in "{fn}", '
+                'supported major version is {major_supp}.'
+            ).format(major=major, fn=filename, major_supp=VERSION_MAJOR)
+        )
+
+    if minor < VERSION_MINOR:
+        raise StoredTransactionVersionError(
+            _('Incompatible minor version {minor} in "{fn}", '
+                'supported minor version is {minor_supp} or greater.'
+            ).format(minor=minor, fn=filename, minor_supp=VERSION_MINOR)
+        )
+
+
+def store_transaction(transaction, filename="transaction.json"):
+    data = {
+        "version": VERSION,
+        "solver_flags": "",  # TODO
+    }
+    rpms = {}
+    groups = {}
+    environments = {}
+
+    for i in transaction.packages():
+        if i.is_package():
+            data_p = rpms.setdefault("%s.%s" % (i.name, i.arch), {})
+            if i._item.isForwardAction():
+                data_p["to"] = i.evr
+                data_p["action"] = i.action_name
+            elif i._item.isBackwardAction():
+                data_p["from"] = i.evr
+                # make sure not to overwrite the forward action if it was written already
+                if "action" not in data_p:
+                    data_p["action"] = i.action_name
+            else:
+                data_p["weird_reason"] = True
+
+            data_p["reason"] = libdnf.transaction.TransactionItemReasonToString(i.reason)
+        elif i.is_group():
+            group = i.get_group()
+
+            data_p = groups.setdefault(group.getGroupId(), {})
+            data_p["action"] = i.action_name
+        elif i.is_environment():
+            env = i.get_environment()
+
+            data_p = environments.setdefault(env.getEnvironmentId(), {})
+            data_p["action"] = i.action_name
```

Comment:
Shall we at least set a warning when additional type of element will be in transaction?

Reply:
I don't understand you here, what is the additional type of element you're talking about?

Reply:
The problem is that when any additional transaction element type is in transaction, the code simply skips it without a warning to a user. Example: In case that transaction will only change a reason of element it simply skips it and silently.

Reply:
Sorry, but I still don't understand. What do you mean by "element" and "element type"?

If you mean there would be and "Install" action for a package with the same version that would only change the reason, I don't think you can generate such action normally? (Installing the same version is a no-op?)